ssm_jax/hmm/core.py

```diff
@@ -0,0 +1,226 @@
+import jax.numpy as jnp
+import jax.random as jr
+from jax import lax
+
+
+def hmm_filter(initial_distribution,
+               transition_matrix,
+               log_likelihoods):
+    """_summary_
+
+    Args:
+        hmm_params (_type_): _description_
+        log_likelihoods (_type_): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    def _step(carry, t):
+        log_normalizer, pred_probs = carry
+
+        # Get parameters for time t
+        get = lambda x, ndim: x[t] if x.ndim == ndim + 1 else x
+        A = get(transition_matrix, 2)
+        ll = log_likelihoods[t]
+
+        # Condition on data at time t, being careful not to overflow.
```

Comment:
How about we factor this logic out into a subroutine like
```
def discrete_update(prior, logliks):
  ll_max = logliks.max()
  post = prior * jnp.exp(logliks - ll_max)
  norm = post.sum()
  post /= norm
  log_normalizer = jnp.log(norm) + ll_max
  return post, log_normalizer
```
and then the main loop becomes
```
filtered_probs, log_normalizer_t = discrete_update(pred_probs, log_likelihoods[t])
log_normalizer_sum += log_normalizer_t
pred_probs = A.T @ filtered_probs
```
Or maybe it should be called `bayes_rule_log` or something like that?

Reply:
How about `_condition_on` and `_predict` like in the `lgssm.core` code?

Reply:
yes!